FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/mqtt_operations.c

```diff
@@ -0,0 +1,1069 @@
+            }
+            else
+            {
+                LogDebug( ( "MQTT connection successfully established with broker." ) );
+            }
+        }
+
+        if( xReturnStatus == true )
+        {
+            /* Keep a flag for indicating if MQTT session is established. This
+             * flag will mark that an MQTT DISCONNECT has to be sent at the end
+             * of the demo even if there are intermediate failures. */
+            xMqttSessionEstablished = true;
+        }
+
+        if( xReturnStatus == true )
+        {
+            /* Check if a session is present and if there are any outgoing
+             * publishes that need to be resent. Resending unacknowledged
+             * publishes is needed only if the broker is re-establishing a
+             * session that was already present. */
+            if( xSessionPresent == true )
+            {
+                LogDebug( ( "An MQTT session with broker is re-established. "
+                            "Resending unacked publishes." ) );
+
+                /* Handle all the resend of publish messages. */
+                xReturnStatus = prvHandlePublishResend( &xMqttContext );
+            }
+            else
+            {
+                LogDebug( ( "A clean MQTT connection is established."
+                            " Cleaning up all the stored outgoing publishes." ) );
+
+                /* Clean up the outgoing publishes waiting for ack as this new
+                 * connection doesn't re-establish an existing session. */
+                prvCleanupOutgoingPublishes();
+            }
+        }
+    }
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+bool xDisconnectMqttSession( void )
+{
+    MQTTStatus_t xMqttStatus = MQTTSuccess;
+    bool xReturnStatus = false;
+    MQTTContext_t * pxMqttContext = &xMqttContext;
+    NetworkContext_t * pxNetworkContext = &xNetworkContext;
+
+    configASSERT( pxMqttContext != NULL );
+    configASSERT( pxNetworkContext != NULL );
+
+    if( xMqttSessionEstablished == true )
+    {
+        /* Send DISCONNECT. */
+        xMqttStatus = MQTT_Disconnect( pxMqttContext );
+
+        if( xMqttStatus != MQTTSuccess )
+        {
+            LogError( ( "Sending MQTT DISCONNECT failed with status=%u.",
+                        xMqttStatus ) );
+        }
+        else
+        {
+            /* MQTT DISCONNECT sent successfully. */
+            xReturnStatus = true;
+        }
+    }
+
+    /* End TLS session, then close TCP connection. */
+    ( void ) TLS_FreeRTOS_Disconnect( pxNetworkContext );
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+bool xSubscribeToTopic( const char * pcTopicFilter,
+                        uint16_t usTopicFilterLength )
+{
+    bool xReturnStatus = false;
+    MQTTStatus_t xMqttStatus;
+    MQTTContext_t * pxMqttContext = &xMqttContext;
```

Comment:
This should be a parameter rather than a global.

Reply:
These mqtt_operations changes will be addressed in a future PR, where mqtt_demo_helpers is adapted to work with PKCS11.